File: 3 Kyu/3 kyu - Make a spiral.c
#include <stdio.h>
#include <stdbool.h>

typedef struct _point { unsigned x, y; } point;
const point direc[4] = {{0,1},{1,0},{0,-1},{-1,0}};

void clear (unsigned n, int grid[n][n]) {
    for (unsigned i = 0; i < n; i++)
        for (unsigned j = 0; j < n; j++)
            grid[i][j] = 0;
}
void spiralize (unsigned n, int grid[n][n]) {

    int pos = 0, end =  n - 1;
    point p = {0,0}, nxt, sup;
}

int main () {

    int size = 15;
    bool grid[15][15] = {{0}};
    int y = 0, x = 0;
    int dir = 0;
    int maxx = size - 1, maxy = size - 1;
    int minx = 0, miny = 0;

    while (1) {
        grid[y][x] = 1;
        
        switch (dir) {

            case 0 : if (x == maxx) { miny += 2, dir = 1; } else x++; break;
            case 1 : if (y == maxy) { maxx -= 2, dir = 2; } else y++; break
            case 2 : if (x == minx) { maxy -= 2, dir = 3; } else x--; break;
            case 3 : if (y == miny) { minx += 2, dir = 0; } else y--; break;

            if (minx > maxx && miny > maxy) break;
        }
    }

    for (int y = 0; y < size; y++) {
        for (int x = 0; x < size; x++) {
            printf ("%i ", grid[y][x]);
        }
        printf ("\n");
    }

}

File: 3 Kyu/3 kyu - Make a spiral.rs
fn spiralize(size: usize) -> Vec<Vec<i8>> {

    let mut grid:Vec<Vec<i8>> = vec![vec![0; size]; size];
    let mut x = 0;
    let mut y = 0;
    let mut dir = 0;
    let mut minx = 0;
    let mut miny = 0;
    let mut maxx = size - 1;
    let mut maxy = size - 1;

    loop {
        grid[y][x] = 1;

        match dir {
            0 => if x == maxx { miny += 2 ; dir = 1 } else { x += 1 },
            1 => if y == maxy { maxx -= 2 ; dir = 2 } else { y += 1 },
            2 => if x == minx { maxy -= 2 ; dir = 3 } else { x -= 1 },
            3 => if y == miny { minx += 2 ; dir = 0 } else { y -= 1 },
            _ => (),
        }
        if minx > maxx && miny > maxy { break }
    }

    grid
}
